### PR TITLE
Fixed #1290 - Query API updates (Alias etc..)

### LINKED ISSUE #1290: 2.0: Query API updates (Alias etc..)
Hello,

I updated the API spec as follows:

Query
1. Meta.documentID -> Meta.id
2. Result implements ReadOnlyArrayInterface and ReadOnlyDictionaryInterface

Type Setter
1. Add type setters to DictionaryInterface, ArrayInterface, and Parameters

Commits:
https://github.com/couchbaselabs/couchbase-lite-apiv2/commit/61445ac38d8fdd98643728604abed03cfe86a99d
https://github.com/couchbaselabs/couchbase-lite-apiv2/commit/d0e3291199958200ed22067a8c2cdd0968f2dc32

— Pasin

### PR DESCRIPTION
**Pull request by hideki**
- Query Alias
- Query Result/ResultSet updates

### CODE REVIEW

File: shared/src/main/java/com/couchbase/lite/CBLFLArray.java [hunk 1]
```diff
@@ -5,11 +5,12 @@
 class CBLFLArray {
     private FLArray flArray;
     private CBLC4Doc c4doc;
+    private CBLFLDataSource lflDataSource;
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/CBLFLArray.java]: should be flDataSource.

File: shared/src/main/java/com/couchbase/lite/CBLFLArray.java [hunk 2]
```diff
@@ -21,6 +22,10 @@ CBLC4Doc getC4doc() {
         return c4doc;
     }
 
+    CBLFLDataSource getLflDataSource() {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/CBLFLArray.java]: should be getFlDataSource()

File: shared/src/main/java/com/couchbase/lite/CBLFLDict.java [hunk 3]
```diff
@@ -4,21 +4,21 @@
 
 class CBLFLDict {
     private FLDict flDict;
-    private CBLC4Doc c4doc;
+    private CBLFLDataSource lflDataSource;
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/CBLFLDict.java]: should be flDataSource.

File: shared/src/main/java/com/couchbase/lite/CBLFLDict.java [hunk 4]
```diff
@@ -4,21 +4,21 @@
...
         this.database = database;
     }
 
     FLDict getFlDict() {
         return flDict;
     }
 
-    CBLC4Doc getC4doc() {
-        return c4doc;
+    CBLFLDataSource getLflDataSource() {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/CBLFLDict.java]: should be getFlDataSource().

File: shared/src/main/java/com/couchbase/lite/Result.java [hunk 5]
```diff
@@ -75,30 +80,25 @@ public long getSequence() {
      * @return the document object.
      */
     public Document getDocument() {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Result.java]: Not in spec.

File: shared/src/main/java/com/couchbase/lite/Result.java [hunk 6]
```diff
@@ -107,146 +107,164 @@ public int count() {
...
     }
 
     @Override
     public Date getDate(int index) {
-        return null;
+        return DateUtils.fromJson(getString(index));
     }
 
     @Override
     public ReadOnlyArrayInterface getArray(int index) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Result.java]: It should return ReadOnlyArray not ReadOnlyArrayInterface. Also IIRC, ReadOnlyArrayInterface is internal for Java.

File: shared/src/main/java/com/couchbase/lite/Result.java [hunk 7]
```diff
@@ -107,146 +107,164 @@ public int count() {
...
-        return null;
+        return (ReadOnlyArrayInterface) fleeceValueToObject(index);
     }
 
     @Override
     public ReadOnlyDictionaryInterface getDictionary(int index) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Result.java]: It should return ReadOnlyDictionary not ReadOnlyDictionaryInterface. Also IIRC, ReadOnlyDictionaryInterface is internal for Java.

File: shared/src/main/java/com/couchbase/lite/Result.java [hunk 8]
```diff
@@ -107,146 +107,164 @@ public int count() {
...
-        return null;
+        return (ReadOnlyDictionaryInterface) fleeceValueToObject(index);
     }
 
     @Override
     public List<Object> toList() {
-        return null;
+        List<Object> array = new ArrayList<>();
+        for (int i = 0; i < count(); i++)
+            array.add(getObject(i));
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Result.java]: toList() should return only JSON object. I understand that getObject(i) can return CBL Array() or CBL Dictionary() ?

File: shared/src/main/java/com/couchbase/lite/Result.java [hunk 9]
```diff
@@ -107,146 +107,164 @@ public int count() {
...
 
     @Override
     public ReadOnlyArrayInterface getArray(String key) {
-        return null;
+        int index = indexForColumnName(key);
+        return index >= 0 ? getArray(index) : null;
     }
 
     @Override
     public ReadOnlyDictionaryInterface getDictionary(String key) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Result.java]: Should return ReadOnlyDictionary.

File: shared/src/main/java/com/couchbase/lite/Result.java [hunk 10]
```diff
@@ -107,146 +107,164 @@ public int count() {
...
 
     @Override
     public Date getDate(String key) {
-        return null;
+        int index = indexForColumnName(key);
+        return index >= 0 ? getDate(index) : null;
     }
 
     @Override
     public ReadOnlyArrayInterface getArray(String key) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Result.java]: Should return ReadOnlyArray.

File: shared/src/main/java/com/couchbase/lite/Result.java [hunk 11]
```diff
@@ -107,146 +107,164 @@ public int count() {
...
+        int index = indexForColumnName(key);
+        return index >= 0 ? getDictionary(index) : null;
     }
 
     @Override
     public Map<String, Object> toMap() {
-        return null;
+        Map<String, Object> dict = new HashMap<>();
+        for (String name : rs.getColumnNames().keySet()) {
+            dict.put(name, getObject(indexForColumnName(name)));
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Result.java]: toMap() should return only JSON based object. I think getObject() can return non JSON based objects.

File: shared/src/main/java/com/couchbase/lite/Result.java (hunk 5 above)
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/Result.java]: `getDocument()`, `getDocumentID()` and `getSequence()` will be kept as a package level method.

**Comment by hideki** [comment]
@pasin, could you please review the PR again?

**Pull request merged**